### ISSUE TITLE
UseDotNet task useGlobalJson finding algorithm totally unusable

### ISSUE THREAD

**Issue by NetTecture** [microsoft/azure-pipelines-tasks#11840]
Current version.

As per the documentation (https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/tool/dotnet-core-tool-installer?view=azure-devops) that is defined as:

* useGlobalJson: true to enable the function
* workingDirectory to specify the path erhe the global.json files should be searched.

SOMEONE seems to have implemented that as searching from the working directory down and then taking a random path. The reality is that as per the global.json file specifications one should start at a specific place (the folder for the solution) and search FROM THERE UP taking the first hit.

My current use case is this:

* We have a repository with multiple solution folders (Source, Source.Backend, Source.Websites etc.)
* There is a global.json in the root folder where the source folders are defined.
* There is in one or two source folders a global.json file that overrides the sdk version for this specific folder.

No problem in Visual Studio. Unusable in Azure.

I have defined a version to find the global.json as this:

- task: UseDotNet@2
  displayName: 'Use .NET Core sdk'
  inputs:
    useGlobalJson: true
    workingDirectory:  '$(Build.SourcesDirectory)/Source.Backend'

The result is:

##[error]Failed to find global.json at and inside path: d:\a\1\s\Source.Backend

Funny thing is: There IS a global.json in the parent directory of that. And global.json is supposed to be searched from the executed solution (which lives in Source.Backend) UPWARD - not downward.

Documentation:

https://docs.microsoft.com/en-us/dotnet/core/tools/global-json

".NET Core SDK looks for a global.json file in the current working directory (which isn't necessarily the same as the project directory) or one of its parent directories."

Whoever implemented this search did not bother reading the documentation and thus it is unusable in the way it is supposed to be used. Which is:

* Compile a solution in a work directlry,
* So use this feature to set the SDK
* Which is either defined globally further up or overridden for specific folders.

For now - sorry - I ahve to roll back this critical feature and stick with an actually working powershell script that was written after reading the specifications of how global.json is supposed to be searched.

**Comment by hiyadav** [comment]
Hey @NetTecture 

What you have observed is true, the installer task searches for all global.json files inside the WorkingDirectory. For all the global.json files found, it will install all the versions specified in them. The intent is, that user only has to run the installer task once, to install all versions required by their sub-projects. Otherwise they might have to run the installer task multiple times from all of the sub-folders they want to perform actions on.

However, when you use DotNetCLIV2 task to run a dotnet command, the behaviour of searching global.json in upward fashion remains intact. And hence the correct global.json will be used by dotnet.exe.

So if you change the working directory to $(System.DefaultWorkingDirectory) in UseDotNetV2 task, it will install all versions defined in each of global.json. So basically, you environment is all set. Now you can use, DotNetCoreCLIV2 task to run command on any project, and the corresponding version in its or its parent's global.json will be picked by dotnet.exe.

**Comment by NetTecture** [comment]
So, basically, it is unsuitable at all to the task of determining which CLI version to use for a specific work directory at all because someone had grand ideas what it should do whatever he thinks makes sense and does not give a crap about the fact that in most case people will use global.json as it is defined in the specs and used by visual studio.

Even today - if you are an enterprise developer -you use Visual Studio, not Visual Studio Code and the CLI. You use global.json somehwere in the solution folder or up to determine the version of the SDK to use because it is a LOT easier than changing all the projects every time a new version is released. You then use a build server and azure task to compile and you want the global.json behavior to match the one that is used in Visual Studio.

I do NOT run DotNetCoreCLIV2 to run any comand on any project. I use global.json to set the sdk version to be used on a visual studio installer task that then runs on about 60 projects in the folder hierarchy doward. If I would rewrite this to the CLI I would have a monster script and not run parallel builds. Works posibly for small website projects, but not for large solutions. Not. At. All.

So, totally useless feature. Gratulations.

Makes me literally wonder about the qualification of the person approving this merge request. This is something I am more used from the EfCore team which is happy to tell me how superiod their product is to EfClassic - while their product produces invalid SQL that throws errors and their timeline for fixing is "put your project on hold some years".

The only good thing is that with an hour work you can use powershell to determine the correct sdk version - the global.json feature basically was designed WITHOUT thinking about the common enterprise use cases.

**Comment by NetTecture** [comment]
@hiyadav Thanks for the explanation. Did you bother thinking about it because what you describe is NOT WHAT HAPPENS?

Here is a use case that is valid, is actually used in larger proejcts and quite obviously something outside of your consideration:. It is exactly OUR use case.

Imagine a larger repository than the typical hello world dotnetcore you seem to consider.

* In the root of the repository there is a global.json because if possible we want to be able to build all with one sdk version and upgrading the project files of 250+ projects is a PAIN.
* Under it are 3 source repositories (Source.A, Source.B, Source.C). Imagine we front end ,backend and additional data load pipelines for a large system.
* Only C redefines a global.json because of some third party library that still is to be ported.

Now, we build in steps, so A, B and C have their own build scripts and are build in parallel in different computers.

Here is your problem. Regardless whether dotnet core cli task or msbuild.... the standard says that you can define a sdk version UPWARDS of the work folder.

* Source.C will work fine becausse your load tasks finds a sdk version.
* Source A and Source B will NOT EVEN SEE THE CENTRALLY DEFINED SDK VERSION. So it is not loaded. While both Msbuild as well as the dotnetcore cli task happily will go upward and find it and then blow with an error that the sdk is not installed.

So, your statement that you install all versions used in a project is WRONG and thus this is a bug. You ONLY install SDK versions that are defined WITHIN a work folder, but those are NOT ALL VERSIONS THAT ARE VALID FOR A FOLDER.

Shitty job and thus we have to stick to a powershell script that is smart enough to go updward because no, you do NOT install all SDK versions at all. YOu happily ignore the centrally defined "use that for all folders wher enothing else is defined" file.

Now sure who did the review on that - but I would question their real world experience.

**Comment by hiyadav** [comment]
Hey @NetTecture 
In above scenario it does look like a bug and I understand the confusion/pain being caused due to the reverse strategy of UseDotNetV2 task's UseGlobalJson feature. The task documentation needs to be more explanatory.

We understand that current design is leading to confusion (like in above scenario), so we will evaluate on adding upwards lookup along with current global.json lookup. (however, it will now work if partial checkout of repository is done and the parent global.json is not checked-out on agent.)

### As per the current design, here is what can be done: 
#### You can use the _**parentFolder**_ containing source.a/b/c as the _**workingDirectory**_ for just **_UseDotNet_** task. If this is done, 
- the task will ensure to get versions from global.json in parentFolder and also in source.c directory. 
it would look like this:
**_task: UseDotNet@2
displayName: 'Use .NET Core sdk'
inputs:
useGlobalJson: true
workingDirectory: '$(Build.SourcesDirectory)'_** (assuming Build.SourcesDirectory contains the global.json to be used by Source.Backend)

Now when you run the scripts for source.a/b/c they should all work as both required versions were downloaded. The let down can be that more SDK's will be installed than needed.

But if source.a/b/c are being built on seperate machine/agents, then UseDotNet task needs to be used in each pipeline and parentFolder can be set as workingDirectory in UseDotNet task. 
The build/test/deploy scripts need not be changed neither the working directory for them needs to be changed. The dotnet cli or msbuild will be able to find the required version.

The current behaviour was designed for setting up environment for entire repository, hence the recommendation of keeping the root folder as working directory for this task.

**Comment by NickDarvey** [comment]
I just stumbled on this as I had to use the `UseDotNet@2` task for the first time to install a recently released version of the SDK. The current behaviour of the task is excellent,

>The intent is, that user only has to run the installer task once, to install all versions required by their sub-projects.

This is exactly what I needed, and after I enabled it with `useGlobalJson: true`, it worked exactly as expected. Thanks for this feature!
